dash/development/_r_components_generation.py

```diff
@@ -0,0 +1,511 @@
+from __future__ import absolute_import
+
+import os
+import sys
+import shutil
+import glob
+
+from ._all_keywords import r_keywords
+from ._py_components_generation import reorder_props
+
+
+# Declaring longer string templates as globals to improve
+# readability, make method logic clearer to anyone inspecting
+# code below
+r_component_string = '''{prefix}{name} <- function(..., {default_argtext}) {{
+
+    component <- list(
+        props = list({default_paramtext}),
+        type = '{name}',
+        namespace = '{project_shortname}',
+        propNames = c({prop_names}),
+        package = '{package_name}'
+        )
+
+    component$props <- filter_null(component$props)
+    component <- append_wildcard_props(component, wildcards = {default_wildcards}, ...)
+
+    structure(component, class = c('dash_component', 'list'))
+}}'''  # noqa:E501
+
+# the following strings represent all the elements in an object
+# of the html_dependency class, which will be propagated by
+# iterating over _js_dist in __init__.py
+frame_open_template = '''.{rpkgname}_js_metadata <- function() {{
+deps_metadata <- list('''
+
+frame_element_template = '''`{dep_name}` = structure(list(name = "{dep_name}",
+version = "{project_ver}", src = list(href = NULL,
+file = "lib/"), meta = NULL,
+script = "{dep_rpp}",
+stylesheet = NULL, head = NULL, attachment = NULL, package = "{rpkgname}",
+all_files = FALSE), class = "html_dependency")'''
+
+frame_body_template = '''`{project_shortname}` = structure(list(name = "{project_shortname}",
+version = "{project_ver}", src = list(href = NULL,
+file = "lib/"), meta = NULL,
+script = "{dep_rpp}",
+stylesheet = NULL, head = NULL, attachment = NULL, package = "{rpkgname}",
+all_files = FALSE), class = "html_dependency")'''
+
+frame_close_template = ''')
+return(deps_metadata)
+}'''
+
+help_string = '''% Auto-generated: do not edit by hand
+\\name{{{prefix}{name}}}
+\\alias{{{prefix}{name}}}
+\\title{{{name} component}}
+\\description{{
+{description}
+}}
+\\usage{{
+{prefix}{name}(..., {default_argtext})
+}}
+\\arguments{{
+{item_text}
+}}
+'''
+
+
+# This is an initial attempt at resolving type inconsistencies
+# between R and JSON.
+def json_to_r_type(current_prop):
+    object_type = current_prop['type'].values()
+    if 'defaultValue' in current_prop and object_type == 'string':
+        if "\"" in current_prop['defaultValue']['value']:
+            argument = current_prop['defaultValue']['value']
+        else:
+            argument = "'{}'".format(current_prop['defaultValue']['value'])
+    elif 'defaultValue' in current_prop and object_type == ['object']:
+        argument = 'list()'
+    elif 'defaultValue' in current_prop and \
+            current_prop['defaultValue']['value'] == '[]':
+        argument = 'list()'
+    else:
+        argument = 'NULL'
+    return argument
+
+
+# pylint: disable=R0914
+def generate_class_string_r(name, props, project_shortname, prefix):
```

Comment:
> I don't think we need the suffix `_r`, since it is clear by the file name that these functions to generate R components, and we do not do an analogous suffix for the Python functions. If we added that suffix, I think we should add it to the Python generation functions too. I don't have a particularly strong opinion for either naming scheme, as long as we are consistent.

I can remove the `_r` suffix, would prefer not to impose my arbitrary naming scheme on anyone else.